lib/sdn_epc/forwarder.ex

```diff
@@ -35,16 +39,50 @@ defmodule SdnEpc.Forwarder do
   end
 
   def handle_call({:open_of_channel, args}, _from, state) do
-    {:ok, _conn_pid} = :ofp_channel.open args[:sup],
+    {:ok, _conn_pid} = :ofp_channel.open(args[:sup],
       "ofp_channel_" <> args[:switch_id],
       {:remote_peer, args[:ip], args[:port], :tcp},
-      controlling_process: __MODULE__, version: args[:version]
+      controlling_process: __MODULE__, version: args[:version])
     {:reply, :ok, state}
   end
 
-  def handle_cast({:handle_switch_msg, datapatch_id: datapatch_id,
+  def handle_cast({:save_datapath_id, datapath_id}, _state) do
+    {:noreply, %{datapath_id: datapath_id}}
+  end
+  def handle_cast({:subscribe_switch_msg, datapath_id: datapath_id,
                   type: type}, state) do
-    :ofs_handler.subscribe datapatch_id, SdnEpc.OfshCalls, type
+    :ofs_handler.subscribe(datapath_id, SdnEpc.OfshCalls, type)
+    {:noreply, state}
+  end
+  def handle_cast({:send_msg_to_controller, switch_id,
+                   msg = {:packet_in, _, _}}, state) do
+    msg_converted = SdnEpc.Converter.ofp_packet_in(msg)
+    :ofp_channel.send(switch_id, msg_converted)
+    Logger.info("Packet in message send to controller")
+    {:noreply, state}
+  end
+  def handle_cast({:send_msg_to_controller, switch_id,
+                   msg = {:features_reply, _, _}}, state) do
+    msg_converted = SdnEpc.Converter.ofp_features_reply(msg)
+    :ofp_channel.send(switch_id, msg_converted)
+    Logger.info("Features reply send to controller")
     {:noreply, state}
   end
+  def handle_cast({:send_msg_to_controller, switch_id,
```

Comment:
It would be much better not to pattern match on every message type in forwarder but do the pattern matching in functions heads in the converter, which would look like:

`Converter.conver(msg)`